clients/venice-consumer/src/main/java/com/linkedin/venice/client/consumer/VeniceChangelogConsumerImpl.java

```diff
@@ -177,30 +226,110 @@ public Collection<PubSubMessage<K, ChangeEvent<V>, Long>> poll(long timeoutInMs)
     return pubSubMessages;
   }
 
+  protected <T> T bufferAndAssembleRecordChangeEvent(
+      PubSubTopicPartition pubSubTopicPartition,
+      int schemaId,
+      byte[] keyBytes,
+      byte[] valueBytes,
+      long recordOffset,
+      AbstractAvroChunkingAdapter<T> chunkingAdapter,
+      Lazy<RecordDeserializer<T>> recordDeserializer,
+      int readerSchemaId,
+      ReadOnlySchemaRepository schemaRepository) {
+    T recordChangeEvent = null;
+    if (storeChunkingEnabled) {
+      if (!inMemoryStorageEngine.containsPartition(pubSubTopicPartition.getPartitionNumber())) {
+        inMemoryStorageEngine.addStoragePartition(pubSubTopicPartition.getPartitionNumber());
+      }
+      if (schemaId == AvroProtocolDefinition.CHUNK.getCurrentProtocolVersion()) {
+        // store in the database and walk away
+        inMemoryStorageEngine.put(
+            pubSubTopicPartition.getPartitionNumber(),
+            keyBytes,
+            ValueRecord.create(schemaId, valueBytes).serialize());
+        return null;
+      } else if (schemaId == AvroProtocolDefinition.CHUNKED_VALUE_MANIFEST.getCurrentProtocolVersion()) {
+        // This is the last value. Store it, and now read it from the in memory store as a fully assembled value
+        inMemoryStorageEngine.put(
+            pubSubTopicPartition.getPartitionNumber(),
+            keyBytes,
+            ValueRecord.create(schemaId, valueBytes).serialize());
+        try {
+          recordChangeEvent = chunkingAdapter.get(
+              inMemoryStorageEngine,
+              readerSchemaId,
+              pubSubTopicPartition.getPartitionNumber(),
+              ByteBuffer.wrap(keyBytes),
+              false,
+              null,
+              null,
+              null,
+              CompressionStrategy.NO_OP,
```

Comment:
It looks kind of whack admittedly, but it is actually correct!  We're configuring this storage engine very specifically. We aren't storing any compressed data here.  We actually decompress it ahead of time.

Reply:
Decompress it ahead of time...? How does that work? On the producing side, we compress, then chunk... so on the consuming side we ought to do the opposite: assemble, then decompress.

Is there something special happening here?

Reply:
Right now we decompress and then assemble on the client side in this patch. Hum...  I know it works but now you make me wonder how 😅

Change capture data in CC today is chunked but never compressed so only vt data is chunked and compressed.

I'll stare at this again Monday.

Reply:
Oh, the CC topic messages are never compressed? Then that code woulf work... But why not compress it? I would expect them to compress very well, given that before/after images may be very similar...

Reply:
It was admittedly an oversight to begin with, but I didn't think to go back and fix it as it has ingestion rate implications.

The way the code in the ingestion task is structured, we do a read on the original value before comparing it with an incoming value.  It's at this point both values are not compressed or chunked.  It's here that we offer up the two values to the view writer to process.  The thinking here is that as we add views down the road, we may need these values in the clear for things like a projection view or some other kind of stored procedure.  If we compressed in the view before writing to Kafka we'd have to compress all messages twice.

I'm still slightly dubious if this is the right thing because this code path also has to assemble chunked messages in the version topic which will be compressed...  But the test case passed so maybe I'm overthinking it

Reply:
Not bad hoss.  This **is** a bug.  But it didn't show up in the testing because there is another bug (that I had patched in a different local branch but no this one).  In the test the behavior is that it instantiates the after image consumer, **which by default starts at the beginning of the change capture topic**, but this behavior is kind of asymmetrical as the non after image consumer starts at the beginning of the version topic.  Ideally both should start at the beginning of the version topic, and then either through api calls or configs, the user should seek to the beginning of RT writes (or the end of RT writes, or a checkpoint or whatever).

If you tweak this, then things start to break.  So I'll patch this.